CupertinoDatePicker & CupertinoTimerPicker dark mode

LongCatIsLooong opened the pull request:
## Description

- Changing the default background color of `CupertinoDatePicker` and `CupertinoTimerPicker` to `null`, which disables background painting entirely (i.e. the picker is going to have a completely transparent background), to match the native `UIPicker` and `UIDatePicker`.
- Add dark mode support.
- Disable the gradient painting when the background is null, because it looks really bad when the background is dark and `backgroundColor = null`

## Golden Changes
https://github.com/flutter/goldens/pull/43/files?short_path=0bfad5e#diff-0bfad5e92f7d3a6380af2f64b62d3cf2

## Related Issues

https://github.com/flutter/flutter/issues/35541
Mitigates https://github.com/flutter/flutter/issues/39685
TODO: https://github.com/flutter/flutter/issues/16831

## Screenshots over black background, with `backgroundColor = null`.
The font color is set to blue so it's visible in both screenshots.

| old | new |
| ---|---|
| ![Screenshot_1567713620](https://user-images.githubusercontent.com/31859944/64376553-66be9980-cfdd-11e9-821d-3e8278f9e9fc.png) | ![Screenshot_1568067257](https://user-images.githubusercontent.com/31859944/64570171-97237200-d314-11e9-8b3e-f1cedf3ca593.png) |

## Tests

I added the following tests:

picker dark mode

Fixed goldens.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

LongCatIsLooong committed:
WIP

LongCatIsLooong committed:
change default color to transparent

LongCatIsLooong committed:
minor changes

LongCatIsLooong committed:
fix tests

LongCatIsLooong committed:
add test

HansMuller commented on packages/flutter/lib/src/cupertino/date_picker.dart:
```diff
@@ -302,7 +299,7 @@ class CupertinoDatePicker extends StatefulWidget {
   /// Background color of date picker.
   ///
-  /// Defaults to [CupertinoColors.white] when null.
+  /// Defaults to null, which disables background painting entirely.
```
Please include this detail in the issue's description, explain that we're making the change to improve fidelity wrt native iOS, and include before/after screenshots to help developers understand what's changed.

LongCatIsLooong replied:
updated

justinmc reviewed (approved):
LGTM :+1:

justinmc commented:
As we roll out these color changes, is the plan to change default background colors everywhere to transparent if they're white?

HansMuller commented on packages/flutter/lib/src/cupertino/picker.dart:
```diff
@@ -331,18 +327,13 @@ class _CupertinoPickerState extends State<CupertinoPicker> {
   @override
   Widget build(BuildContext context) {
-    Widget result = DefaultTextStyle(
+    final Color resolvedBackgroundColor = widget.backgroundColor == null
```
If this idiom crops up enough, maybe we should allow CupertinoDyanmicColor.resolve(null, context)?

LongCatIsLooong replied:
https://github.com/flutter/flutter/pull/39927

justinmc replied:
Anything to change now that that PR has been merged?

HansMuller reviewed (approved):
LGTM

LongCatIsLooong committed:
review

LongCatIsLooong commented:
> As we roll out these color changes, is the plan to change default background colors everywhere to transparent if they're white?

The main reason the picker is made transparent in this PR is that the native iOS `UIPicker` has a transparent background color. I don't think it applies to widgets like `CupertinoDialog` despite having a white background.

LongCatIsLooong committed:
Opacity

LongCatIsLooong committed:
update tests

LongCatIsLooong commented:
@HansMuller @justinmc updated `ListWheelScrollView`, removed the gradients and replaced it with `offCenterOpacity`. It's a big change, could you re-review this PR?

LongCatIsLooong committed:
update

LongCatIsLooong commented:
It includes changes from https://github.com/flutter/flutter/pull/33070, in order for the opacity layer to work  properly.

HansMuller commented:
The new changes look good to me; just had a few small comments.

justinmc commented on packages/flutter/lib/src/cupertino/picker.dart:
```diff
@@ -239,110 +243,38 @@ class _CupertinoPickerState extends State<CupertinoPicker> {
-              color: foreground,
-            ),
-          ),
+          Expanded(child: Container()),
```
Do you still need the container now that there is no color?  Or, could the Expanded be replaced with a `Spacer`?  Also line 265 below.

I'm not sure, just throwing some ideas out there :)

LongCatIsLooong replied:
Yeah makes sense. I'm going to just center the middle `Container`

justinmc commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -368,6 +372,26 @@ class RenderListWheelViewport
+  double _offCenterOpacity = 1.0;
+  set offCenterOpacity(double value) {
+    assert(value != null);
+    assert(value > 0);
```
Should this be `value >= 0`?  Looking at your comment.

LongCatIsLooong replied:
ah forgot to change it. nice catch!

justinmc reviewed (approved):
LGTM

The only important thing to check is a potentially incorrect `>` commented above.

LongCatIsLooong committed:
Merge branch 'master' into dark-mode-d

LongCatIsLooong committed:
comments

LongCatIsLooong committed:
update

Piinks commented on bin/internal/goldens.version:
```diff
@@ -1 +1 @@
-49f8198b72f6e12c65fe1db2e46162de0204e671
+38e117bcd79860a3f239872c579acac716e70d49
```
This commit does not appear to exist in https://github.com/flutter/goldens/commits/master which is causing these golden file tests to fail.

LongCatIsLooong replied:
Ah sorry I haven't merged the golden PR yet. I changed the golden hash to prevent `flutter test` from resetting the head to `master` so I can test with my golden fork. I'll update it once the golden PR is merged.

LongCatIsLooong replied:
Updated.

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into dark-mode-d

LongCatIsLooong committed:
update golden hash

LongCatIsLooong committed:
Merge branch 'dark-mode-d' of github.com:LongCatIsLooong/flutter into dark-mode-d

LongCatIsLooong committed:
review

Piinks reviewed (approved):
Goldens look good to me! Thanks! The bot may notify you about triage on Flutter Gold when this lands. Let me know if you have any questions. :)

LongCatIsLooong commented:
https://github.com/flutter/flutter/commit/9c6dd0de231d2f16352419da6a193c2f9dbad31c seems to cause the golden test to fail

dannyvalentesonos commented:
@LongCatIsLooong 
In your description, you wrote:
```
Disable the gradient painting when the background is null, because it looks really bad when the background is dark and backgroundColor = null
```
which to me means that gradient painting is only disabled when background is null, however, it seems like you removed the gradient entirely, and it's now impossible to get the same gradient effect as before. Am I right? 

If so, this causes a problem for me and how I was using this widget. Is there a reason for that?

Thanks!

LongCatIsLooong commented:
@dannyvalentesonos yes the gradient was removed entirely. The gradient has hardcoded colors that didn't work well with a dark background so it was removed.
